pkg/operator/handlers/dapr_handler.go

```diff
@@ -51,6 +51,11 @@ type DaprHandler struct {
 	Scheme *runtime.Scheme
 }
 
+type innerReconciler struct {
```

Comment:
I am hesitate that a `innerReconciler` is necessary, and I believe it is a good idea to keep Handler interface having `Reconcile` method. I make a quick change to verify my idea, pls. check below:

```diff
diff --git a/pkg/operator/handlers/dapr_handler.go b/pkg/operator/handlers/dapr_handler.go
index 19a5456d..16125ae7 100644
--- a/pkg/operator/handlers/dapr_handler.go
+++ b/pkg/operator/handlers/dapr_handler.go
@@ -51,11 +51,6 @@ type DaprHandler struct {
 	Scheme *runtime.Scheme
 }
 
-type innerReconciler struct {
-	h       *DaprHandler
-	factory func() client.Object
-}
-
 // NewDaprHandler returns a new Dapr handler.
 func NewDaprHandler(mgr ctrl.Manager) *DaprHandler {
 	return &DaprHandler{
@@ -75,7 +70,8 @@ func (h *DaprHandler) Init() error {
 		func(rawObj client.Object) []string {
 			svc := rawObj.(*corev1.Service)
 			owner := meta_v1.GetControllerOf(svc)
-			if owner == nil || owner.APIVersion != appsv1.SchemeGroupVersion.String() || (owner.Kind != "Deployment" && owner.Kind != "StatefulSet") {
+			if owner == nil || owner.APIVersion != appsv1.SchemeGroupVersion.String() ||
+				(owner.Kind != "Deployment" && owner.Kind != "StatefulSet") {
 				return nil
 			}
 			return []string{owner.Name}
@@ -83,57 +79,57 @@ func (h *DaprHandler) Init() error {
 		return err
 	}
 
-	if err := ctrl.NewControllerManagedBy(h.mgr).
+	err := ctrl.NewControllerManagedBy(h.mgr).
 		For(&appsv1.Deployment{}).
 		Owns(&corev1.Service{}).
-		WithOptions(controller.Options{
-			MaxConcurrentReconciles: 100,
-		}).
-		Complete(&innerReconciler{
-			h: h,
-			factory: func() client.Object {
-				return &appsv1.Deployment{}
-			},
-		}); err != nil {
+		WithOptions(controller.Options{MaxConcurrentReconciles: 100}).
+		Complete(h)
+
+	if err != nil {
 		return err
 	}
+
 	return ctrl.NewControllerManagedBy(h.mgr).
 		For(&appsv1.StatefulSet{}).
 		Owns(&corev1.Service{}).
-		WithOptions(controller.Options{
-			MaxConcurrentReconciles: 100,
-		}).
-		Complete(&innerReconciler{
-			h: h,
-			factory: func() client.Object {
-				return &appsv1.StatefulSet{}
-			},
-		})
+		WithOptions(controller.Options{MaxConcurrentReconciles: 100}).
+		Complete(h)
 }
 
 // Reconcile the expected services for deployments | statefulset annotated for Dapr.
-func (i *innerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (h *DaprHandler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	// var obj appsv1.Deployment | appsv1.StatefulSet
-	obj := i.factory()
+	result, err := h.reconcile(ctx, req, &appsv1.Deployment{})
+	if err != nil {
+		if result.Requeue {
+			return result, err
+		} else {
+			result, err = h.reconcile(ctx, req, &appsv1.StatefulSet{})
+		}
+	}
+	return result, err
+}
 
+func (h *DaprHandler) reconcile(ctx context.Context, req ctrl.Request, obj client.Object) (ctrl.Result, error) {
 	expectedService := false
-	if err := i.h.Get(ctx, req.NamespacedName, obj); err != nil {
+	kind := obj.GetObjectKind().GroupVersionKind().Kind
+	if err := h.Get(ctx, req.NamespacedName, obj); err != nil {
 		if apierrors.IsNotFound(err) {
-			log.Debugf("deployment has be deleted, %s", req.NamespacedName)
+			log.Debugf("%s has be deleted, %s", kind, req.NamespacedName)
 		} else {
-			log.Errorf("unable to get deployment, %s, err: %s", req.NamespacedName, err)
+			log.Errorf("unable to get %s, %s, err: %s", kind, req.NamespacedName, err)
 			return ctrl.Result{}, err
 		}
 	} else {
 		if obj.GetDeletionTimestamp() != nil {
-			log.Debugf("deployment is being deleted, %s", req.NamespacedName)
+			log.Debugf("%s is being deleted, %s", kind, req.NamespacedName)
 			return ctrl.Result{}, nil
 		}
-		expectedService = i.h.isAnnotatedForDapr(obj)
+		expectedService = h.isAnnotatedForDapr(obj)
 	}
 
 	if expectedService {
-		if err := i.h.ensureDaprServicePresent(ctx, req.Namespace, obj); err != nil {
+		if err := h.ensureDaprServicePresent(ctx, req.Namespace, obj); err != nil {
 			return ctrl.Result{Requeue: true}, err
 		}
 	}
diff --git a/pkg/operator/handlers/handler.go b/pkg/operator/handlers/handler.go
index 5a180ee0..55384d68 100644
--- a/pkg/operator/handlers/handler.go
+++ b/pkg/operator/handlers/handler.go
@@ -1,6 +1,12 @@
 package handlers
 
+import (
+	"context"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
 // Handler is the interface for dealing with Dapr CRDs state changes.
 type Handler interface {
 	Init() error
+	Reconcile(context.Context, ctrl.Request) (ctrl.Result, error)
 }

```

Reply:
one more thought, if we introduce a new interface like this:

```go
type DaprObject interface {
	GetLabels() map[string]string
	GetAnnotations() map[string]string
}
```

then wrap `appsv1.Deployment` and `appsv1.StatefulSet` to implement it, it would be better:

```go
type deploymentWrapper struct {
	deployment *appsv1.Deployment
}

func (d *deploymentWrapper) GetLabels() map[string]string {
	return d.deployment.Spec.Selector.MatchLabels
}

func (d *deploymentWrapper) GetAnnotations() map[string]string {
	return d.deployment.Spec.Template.ObjectMeta.Annotations
}
```

Reply:
I agree with @beiwei30. @theonefx please make suggested changes.

Reply:
ping @theonefx

Reply:
done